pkg/agent/route/route_windows.go

```diff
@@ -0,0 +1,155 @@
+	return &Client{
+		nr:          nr,
+		serviceCIDR: serviceCIDR,
+		hostRoutes:  &sync.Map{},
+	}, nil
+}
+
+// Initialize sets nodeConfig on Window.
+// Service LoadBalancing is provided by OpenFlow.
+func (c *Client) Initialize(nodeConfig *config.NodeConfig) error {
+	c.nodeConfig = nodeConfig
+	return nil
+}
+
+// Reconcile removes the orphaned routes and related configuration based on the desired podCIDRs. Only the route
+// entries on the host gateway interface are stored in the cache.
+func (c *Client) Reconcile(podCIDRs []string) error {
+	desiredPodCIDRs := sets.NewString(podCIDRs...)
+	routes, err := c.listRoutes()
+	if err != nil {
+		return err
+	}
+	gwLinkIndex := c.nodeConfig.GatewayConfig.LinkIndex
+	for dst, rt := range routes {
+		if rt.LinkIndex != gwLinkIndex {
+			continue
+		}
+		if desiredPodCIDRs.Has(dst) {
+			c.hostRoutes.Store(dst, rt)
+			continue
+		}
+		err := c.deleteRoute(rt)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// AddRoutes adds routes to the provided podCIDR.
+// It overrides the routes if they already exist, without error.
+func (c *Client) AddRoutes(podCIDR *net.IPNet, peerNodeIP, peerGwIP net.IP) error {
+	obj, found := c.hostRoutes.Load(podCIDR.String())
+	if found {
+		rt := obj.(*netroute.Route)
+		if rt.GatewayAddress.Equal(peerGwIP) {
+			klog.V(4).Infof("Route with destination %s already exists", podCIDR.String())
+			return nil
+		}
+		// Remove the existing route entry if the gateway address is not as expected.
+		if err := c.deleteRoute(rt); err != nil {
+			klog.Errorf("Failed to delete existing route entry with destination %s gateway %s", podCIDR.String(), peerGwIP.String())
+			return err
+		}
+	}
+	if err := c.nr.NewNetRoute(c.nodeConfig.GatewayConfig.LinkIndex, podCIDR, peerGwIP); err != nil {
+		return err
+	}
+	c.hostRoutes.Store(podCIDR.String(), &netroute.Route{
+		LinkIndex:         c.nodeConfig.GatewayConfig.LinkIndex,
+		DestinationSubnet: podCIDR,
+		GatewayAddress:    peerGwIP,
+	})
+	klog.V(2).Infof("Added route with destination %s gateway%s on host gateway", podCIDR.String(), peerGwIP.String())
+	return nil
+}
+
+// DeleteRoutes deletes routes to the provided podCIDR.
+// It does nothing if the routes don't exist, without error.
+func (c *Client) DeleteRoutes(podCIDR *net.IPNet) error {
+	obj, found := c.hostRoutes.Load(podCIDR.String())
+	if !found {
+		klog.V(2).Infof("Route with destination %s not exists", podCIDR.String())
+		return nil
+	}
+
+	rt := obj.(*netroute.Route)
+	if err := c.deleteRoute(rt); err != nil {
+		return err
+	}
+	c.hostRoutes.Delete(podCIDR.String())
+	klog.V(2).Infof("Deleted route with destination %s from host gateway", podCIDR.String())
+	return nil
+}
+
+// deleteRoute removes the route entry from the host. Loopback route is skipped when deleting routes.
+func (c *Client) deleteRoute(rt *netroute.Route) error {
+	// Deleting Loopback route is forbidden on Windows, so skip such routes.
+	if rt.DestinationSubnet.IP.IsLoopback() {
```

Comment:
Out of curiosity, why would it try to remove routes to loopback address?

Reply:
The function `deleteRoute` might be called in a loop, and the netroute.Route in the loop is possibly a loopback address. This is because netroute returns all route entries, including an entry using the loopback address as the destinations.

Reply:
I see you have excluded loopback address in `listRoutes` and `addRoutes` only add routes to podCIDR, do I miss anything?

Reply:
Correct. I forgot having added filters in `listRoutes`. Will remove the check here.